### PR TITLE
Streaming Sortformer Diarizer: model and module code [PR 2]

### PR DESCRIPTION
**Pull request by tango4j**
# What does this PR do ?

Adding streaming Sortformer diarizer model code to NeMo main.

This PR can be only merged after [FastConformer -bypass pre-encode PR  #13147](https://github.com/NVIDIA/NeMo/pull/13147).

This PR is adding modules, models code and inference code (e2e_diarizee_speech.py).

**Collection**: 

ASR - Speaker Tasks

# Changelog

`nemo/collections/asr/modules/sortformer_modules.py`
`nemo/collections/asr/models/sortformer_diar_models.py`
`examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py`

# Usage

```python
python $NEMO_BASEPATH/neural_diarizer/e2e_diarize_speech.py \
    batch_size=$BATCH_SIZE \
    streaming_mode=true \
    model_path=${MODEL_PATH} \
    dataset_manifest=${TEST_MANIFEST} 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.
The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

## Who can review?

Anyone in NeMo ASR

# Additional Information

There will be at least 2 more PRs for unit-tests and documents

### CODE REVIEW

File: nemo/collections/asr/models/sortformer_diar_models.py [hunk 1]
```diff
@@ -430,21 +553,303 @@
...
             )
             preds = self.forward_infer(emb_seq, emb_seq_length)
         return preds
 
+    @property
+    def input_names(self):
+        return ["chunk", "chunk_lengths", "spkcache", "spkcache_lengths", "fifo", "fifo_lengths"]
+
+    @property
+    def output_names(self):
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:576]: fixed

File: nemo/collections/asr/models/sortformer_diar_models.py [hunk 2]
```diff
@@ -430,21 +553,303 @@
...
+            self.transformer_encoder.diag = None
+
+        del processed_signal, processed_signal_length
+
+        if self.pad_front:
+            total_offset = processed_signal_offset // self.sortformer_modules.subsampling_factor
+            total_preds = self.shift_emb(total_preds, total_offset)
+
+        if T < max_T:  # discard preds corresponding to padding
+            n_frames = math.ceil(T / self.encoder.subsampling_factor)
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:754]: fixed

**Comment by chtruong814** [comment]
@tango4j I'm re-ordering the CI queue a bit to let some priority release items go through.  Will re-add this PR to the queue later.

File: nemo/collections/asr/models/sortformer_diar_models.py [hunk 3]
```diff
@@ -109,14 +149,19 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
         self.loss = instantiate(self._cfg.loss)
 
+        self.async_streaming = self._cfg.get("async_streaming", False)
         self.streaming_mode = self._cfg.get("streaming_mode", False)
         self.save_hyperparameters("cfg")
         self._init_eval_metrics()
         speaker_inds = list(range(self._cfg.max_num_of_spks))
         self.speaker_permutations = torch.tensor(list(itertools.permutations(speaker_inds)))  # Get all permutations
 
+        self.max_batch_dur = self._cfg.get("max_batch_dur", 20000)
```
Comment (KunalDhawan) [inline nemo/collections/asr/models/sortformer_diar_models.py:162]: Could you share more details on how you chose the default value for max_batch_dur please? This would be GPU memory dependent right? Would 20000 be low enough to not cause OOM in line 510 on most of the supported GPUs?

File: nemo/collections/asr/models/sortformer_diar_models.py [hunk 4]
```diff
@@ -430,21 +531,280 @@ def forward(
...
+                self.negative_init_val,
+                dtype=processed_signal.dtype,
+                device=processed_signal.device,
+            )
+            processed_signal = torch.cat([processed_signal, pad_tensor], dim=2)
+
+        att_mod = False
+        if self.training:
+            r = random.random()
+            if r < self.sortformer_modules.causal_attn_rate:
```
Comment (KunalDhawan) [inline nemo/collections/asr/models/sortformer_diar_models.py:675]: Adding a quick comment here explaining causal_attn_rate, causal_attn_rc and why this is required would be beneficial

File: nemo/collections/asr/modules/sortformer_modules.py [hunk 5]
```diff
@@ -12,68 +12,145 @@
...
 import torch.nn.functional as F
 
 from nemo.core.classes.exportable import Exportable
 from nemo.core.classes.module import NeuralModule
+from nemo.utils import logging
 
 __all__ = ['SortformerModules']
 
 
+def concat_and_pad(embs: List[torch.Tensor], lengths: List[torch.Tensor]):
```
Comment (KunalDhawan) [inline nemo/collections/asr/modules/sortformer_modules.py:30]: There is another function with similar functionality and same name in models/sortformer_diar_models.py, is it possible to combine them and make this function reusable for other tasks?

File: nemo/collections/asr/modules/sortformer_modules.py [hunk 6]
```diff
@@ -12,68 +12,145 @@
...
         num_spks: int = 4,
-        hidden_size: int = 192,
         dropout_rate: float = 0.5,
         fc_d_model: int = 512,
         tf_d_model: int = 192,
+        subsampling_factor: int = 8,
+        spkcache_len: int = 188,
+        fifo_len: int = 0,
+        chunk_len: int = 376,
+        spkcache_refresh_rate: int = 1,
```
Comment (KunalDhawan) [inline nemo/collections/asr/modules/sortformer_modules.py:99]: Would these variables (e.g. spkcache_refresh_rate, spkcache_sil_frames_per_spk, causal_attn_rate, scores_boost_latest, min_pos_scores_rate etc) be explained in the config? A description for them would be helpful for users to correctly set these values

**Review by KunalDhawan — APPROVED** [review]
Thanks, great work! LGTM, left minor comments

File: nemo/collections/asr/modules/sortformer_modules.py (hunk 5 above)
Comment (tango4j) [reply nemo/collections/asr/modules/sortformer_modules.py:30]: This is removed.

File: nemo/collections/asr/models/sortformer_diar_models.py [hunk 7]
```diff
@@ -430,21 +531,280 @@ def forward(
...
             preds = self.forward_infer(emb_seq, emb_seq_length)
         return preds
 
+    @property
+    def input_names(self):
+        return ["chunk", "chunk_lengths", "spkcache", "spkcache_lengths", "fifo", "fifo_lengths"]
+
+    @property
+    def output_names(self):
+        return ["preds"]
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/sortformer_diar_models.py]: It seems like the `input_names` is the same as the arguments of `forward_for_export` function but the `output_names` is the same as `forward` function, not `forward_for_export` function

File: nemo/collections/asr/models/sortformer_diar_models.py (hunk 3 above)
Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:162]: This needs to be revisited when we upload test and examples codes

File: nemo/collections/asr/models/sortformer_diar_models.py (hunk 4 above)
Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py:678]: This needs to be revisited when we add documentations and example codes

File: nemo/collections/asr/modules/sortformer_modules.py [hunk 8]
```diff
@@ -84,19 +161,616 @@ def length_to_mask(self, lengths, max_length):
...
         Returns:
-            preds (torch.Tensor): tensor of shape (batch_size, seq_len, num_spks) containing speaker probabilities
+            preds (torch.Tensor): tensor containing speaker probabilities in Sigmoid activation function
+                Dimension: (batch_size, n_frames, n_spk)
         """
         hidden_out = self.dropout(F.relu(hidden_out))
         hidden_out = self.first_hidden_to_hidden(hidden_out)
         hidden_out = self.dropout(F.relu(hidden_out))
         spk_preds = self.single_hidden_to_spks(hidden_out)
         preds = nn.Sigmoid()(spk_preds)
```
Comment (weiqingw4ng) [inline nemo/collections/asr/modules/sortformer_modules.py:211]: If using sigmoid function it is better to use `nn.functional.sigmoid(spk_preds)`

File: nemo/collections/asr/models/sortformer_diar_models.py (hunk 7 above)
Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py]: updated this.

File: nemo/collections/asr/models/sortformer_diar_models.py [hunk 9]
```diff
@@ -399,17 +492,25 @@ def process_signal(self, audio_signal, audio_signal_length):
...
+            processed_signal, processed_signal_length = self.oom_safe_feature_extraction(
+                input_signal=audio_signal, input_signal_length=audio_signal_length
+            )
+        else:
+            processed_signal, processed_signal_length = self.preprocessor(
+                input_signal=audio_signal, length=audio_signal_length
+            )
+        del audio_signal, audio_signal_length
         if not self.training:
             torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/sortformer_diar_models.py]: added `self.streaming_mode`

File: nemo/collections/asr/modules/sortformer_modules.py (hunk 8 above)
Comment (tango4j) [reply nemo/collections/asr/modules/sortformer_modules.py:211]: @ipmedenn 
Please comment on this. Need to check for this change.

**Review by weiqingw4ng — APPROVED** [review]

**Review by weiqingw4ng — DISMISSED** [review]
LGTM.

File: nemo/collections/asr/modules/sortformer_modules.py (hunk 8 above)
Comment (ipmedenn) [reply nemo/collections/asr/modules/sortformer_modules.py]: @tango4j This is legacy code, I fixed as @weiqingw4ng suggests. Thanks!

**Review by weiqingw4ng — APPROVED** [review]
LGTM.

**Review by ko3n1g — APPROVED** [review]
Good from CI side, please confirm when its ready for merge

**Comment by tango4j** [comment]
> Good from CI side, please confirm when its ready for merge

Thanks @ko3n1g , now it can be merged.

**Comment by rbgreenway** [comment]
Any idea when the SortformerDiarizer that has this streaming mode will be available in NeMo?  Doesn't appear to be included in the latest.

**Comment by KunalDhawan** [comment]
Hi @rbgreenway, thank you for the interest in streaming Sortformer. The model + documentation would be released soon, mostly within the next month.
